File: packages/java/src/test/java/io/github/yauhenbichel/regexparity/RegexParityTest.java
package io.github.yauhenbichel.regexparity;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.github.yauhenbichel.regexparity.RegexParity.Finding;
import io.github.yauhenbichel.regexparity.RegexParity.Match;
import io.github.yauhenbichel.regexparity.RegexParity.PatternException;
import io.github.yauhenbichel.regexparity.RegexParity.Rule;
import io.github.yauhenbichel.regexparity.RegexParity.Variant;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.stream.Collectors;
import org.junit.jupiter.api.Test;

/** The Java package against its own behaviour, and against conformance/cases.tsv. */
class RegexParityTest {

  private static final Path CONFORMANCE = Path.of("..", "..", "conformance");

  private static String cp(int... codePoints) {
    StringBuilder out = new StringBuilder();
    for (int codePoint : codePoints) {
      out.appendCodePoint(codePoint);
    }
    return out.toString();
  }

  @Test
  void aLongSDoesNotGetPastARule() {
    assertTrue(RegexParity.matches("\\bharmless\\b", "That is definitely harmle" + cp(0x017F) + "s.", false));
  }

  @Test
  void findAllReportsRangesOfTheOriginalText() {
    String text = "a mela" + cp(0x200B) + "noma, order " + cp(0xFF11, 0xFF12, 0xFF13);
    assertEquals(List.of(new Match(2, 11, "mela" + cp(0x200B) + "noma")), RegexParity.findAll("\\bmelanoma\\b", text, false));
    assertEquals(cp(0xFF11, 0xFF12, 0xFF13), RegexParity.findAll("\\d{3}", text, false).get(0).text());
  }

  @Test
  void caseSensitivityIsAnOptionAndAsciiOnlyEitherWay() {
    assertTrue(RegexParity.matches("\\bTODO\\b", "todo", false));
    assertFalse(RegexParity.matches("\\bTODO\\b", "todo", true));
    assertFalse(RegexParity.matches("\\bthis\\b", "th" + cp(0x0131) + "s", false), "a dotless i is not an i");
  }

  @Test
  void anAccentedLetterIsNotAWordCharacter() {
    assertTrue(RegexParity.matches("\\bcaf\\b", "caf" + cp(0x00E9), false));
    assertFalse(RegexParity.matches("^\\w+\\s", "caf" + cp(0x00E9) + " ok", false));
  }

  @Test
  void unlessOnlyCountsInsideTheSameSentence() {
    Rule rule = new Rule("date", List.of("\\bready\\s+by\\s+friday\\b"), List.of("\\bif\\b"), false);
    assertEquals(0, RegexParity.evaluate(rule, "Ready by Friday if it passes.").size());
    assertEquals(1, RegexParity.evaluate(rule, "Ready by Friday. If you ask me, late.").size());
  }

  @Test
  void aRangeFoundByTwoPatternsCountsOnce() {
    assertEquals(1, RegexParity.evaluate(new Rule("twice", List.of("\\bharmless\\b", "harmless")), "harmless").size());
  }

  @Test
  void foldingKeepsAMapBackToTheOriginalText() {
    RegexParity.Folded folded = RegexParity.fold("it" + cp(0x2019) + "s" + cp(0x000D, 0x2028) + "ok " + cp(0x2014) + " fine");
    assertEquals("it's\n\nok - fine", folded.text());
    assertEquals(folded.text().length(), folded.starts().length);
  }

  @Test
  void theSentenceFunctionIncludesItsTerminator() {
    assertEquals(List.of(5, 14), List.of(RegexParity.sentenceRange("One. Two here! Three", 6)[0], RegexParity.sentenceRange("One. Two here! Three", 6)[1]));
    assertEquals(List.of(15, 20), List.of(RegexParity.sentenceRange("One. Two here! Three", 16)[0], RegexParity.sentenceRange("One. Two here! Three", 16)[1]));
  }

  @Test
  void patternsEnginesReadDifferentlyAreRefused() {
    String backslash = String.valueOf((char) 92);
    List<String> refused = List.of("(?=a)", "(?<!a)b", "(a)\\1", "(?<name>a)", "(?i)a", "a$", "\\p{L}", "a++", "a{,3}", "x{",
        "[[a]]", "[a&&b]", "[]a]", "\\x41", backslash + "u0041", "\\Aa", "\\v", "caf" + cp(0x00E9), "[\\b]", "a{1001}", "a{2,5000}");
    for (String source : refused) {
      assertFalse(RegexParity.checkPattern(source).isEmpty(), source);
    }
    List<String> accepted = List.of("\\bcolou?r\\b", "a{2,3}", "a{2}", "[a-z\\]]", "\\(\\?=", "(?:ab)+?", "\\d+\\.\\d*", "[^.!?]{0,30}", "\\$5", "a{1000}");
    for (String source : accepted) {
      assertEquals(List.of(), RegexParity.checkPattern(source), source);
    }
    assertThrows(PatternException.class, () -> RegexParity.compile("(?=a)", false));
    assertThrows(PatternException.class, () -> RegexParity.compile("(a", false));
  }

  @Test
  void everyVariantFoldsBackToTheOriginal() {
    String original = "It's \"fine\" - order #123456 kept";
    List<Variant> variants = RegexParity.variants(original, false);
    assertEquals(List.of("long s", "Kelvin sign", "full-width", "zero-width spaces", "no-break spaces", "en dashes", "smart quotes"),
        variants.stream().map(Variant::name).toList());
    for (Variant variant : variants) {
      // The Kelvin sign folds to a capital K, so compare without case.
      assertEquals(RegexParity.fold(original).text().toLowerCase(Locale.ROOT), RegexParity.fold(variant.text()).text().toLowerCase(Locale.ROOT), variant.name());
    }
  }

  // --- conformance: the same answers as conformance/cases.tsv, which JavaScript and Python also read ---

  private static String unescape(String text) {
    StringBuilder out = new StringBuilder();
    for (int i = 0; i < text.length(); i++) {
      char c = text.charAt(i);
      if (c == (char) 92 && i + 5 < text.length() && text.charAt(i + 1) == 'u') {
        out.append((char) Integer.parseInt(text.substring(i + 2, i + 6), 16));
        i += 5;
      } else {
        out.append(c);
      }
    }
    return out.toString();
  }

  private static List<String[]> rows(String name) throws IOException {
    return Files.readAllLines(CONFORMANCE.resolve(name), StandardCharsets.UTF_8).stream()
        .filter(line -> !line.isEmpty() && !line.startsWith("#"))
        .map(line -> line.split("\t", -1))
        .toList();
  }

  private static Map<String, Rule> rules() throws IOException {
    Map<String, List<String>> patterns = new LinkedHashMap<>();
    Map<String, List<String>> unless = new LinkedHashMap<>();
    Map<String, Boolean> sensitive = new LinkedHashMap<>();
    for (String[] row : rows("rules.tsv")) {
      patterns.computeIfAbsent(row[0], k -> new ArrayList<>());
      unless.computeIfAbsent(row[0], k -> new ArrayList<>());
      sensitive.put(row[0], row[1].equals("sensitive"));
      (row[2].equals("pattern") ? patterns : unless).get(row[0]).add(unescape(row[3]));
    }
    Map<String, Rule> out = new LinkedHashMap<>();
    for (String id : patterns.keySet()) {
      out.put(id, new Rule(id, patterns.get(id), unless.get(id), sensitive.get(id)));
    }
    return out;
  }

  @Test
  void conformanceEveryCaseGetsTheRecordedAnswerAndFindings() throws IOException {
    Map<String, Rule> rules = rules();
    List<String[]> cases = rows("cases.tsv");
    assertTrue(cases.size() > 100, "only " + cases.size() + " cases");
    for (String[] row : cases) {
      List<Finding> found = RegexParity.evaluate(rules.get(row[0]), unescape(row[3]));
      String label = row[0] + " " + row[1] + " (" + row[2] + ")";
      assertEquals(row[1].equals("match"), !found.isEmpty(), label);
      String spans = found.stream().map(f -> f.start() + ":" + f.end()).collect(Collectors.joining(","));
      assertEquals(row[4], spans.isEmpty() ? "-" : spans, label);
    }
  }

  @Test
  void conformanceVariantsAreGeneratedExactlyAsRecorded() throws IOException {
    Map<String, Rule> rules = rules();
    String[] base = null;
    List<Variant> expected = new ArrayList<>();
    for (String[] row : rows("cases.tsv")) {
      if (row[2].equals("as written")) {
        if (base != null) {
          assertEquals(expected, RegexParity.variants(unescape(base[3]), rules.get(base[0]).caseSensitive()), base[3]);
        }
        base = row;
        expected = new ArrayList<>();
      } else {
        expected.add(new Variant(row[2], unescape(row[3])));
      }
    }
    assertEquals(expected, RegexParity.variants(unescape(base[3]), rules.get(base[0]).caseSensitive()), base[3]);
  }

  @Test
  void conformanceEveryLanguageRefusesAndAcceptsTheSamePatterns() throws IOException {
    List<String[]> patterns = rows("patterns.tsv");
    assertTrue(patterns.size() > 10);
    for (String[] row : patterns) {
      String source = unescape(row[1]);
      assertEquals(row[0].equals("refused"), !RegexParity.checkPattern(source).isEmpty(), source);
    }
  }
}

File: go/regexparity_test.go
package regexparity

import (
	"os"
	"path/filepath"
	"reflect"
	"strconv"
	"strings"
	"testing"
	"unicode/utf8"
)

func cp(codePoints ...rune) string { return string(codePoints) }

func mustMatch(t *testing.T, pattern, text string, caseSensitive bool) bool {
	t.Helper()
	ok, err := Matches(pattern, text, caseSensitive)
	if err != nil {
		t.Fatalf("%q: %v", pattern, err)
	}
	return ok
}

func TestLongSDoesNotGetPastARule(t *testing.T) {
	if !mustMatch(t, `\bharmless\b`, "That is definitely harmle"+cp(0x017F)+"s.", false) {
		t.Fatal("a long s got past the rule")
	}
}

func TestFindAllReportsRangesOfTheOriginalText(t *testing.T) {
	text := "a mela" + cp(0x200B) + "noma, order " + cp(0xFF11, 0xFF12, 0xFF13)
	found, err := FindAll(`\bmelanoma\b`, text, false)
	if err != nil || len(found) != 1 || found[0].Text != "mela"+cp(0x200B)+"noma" || text[found[0].Start:found[0].End] != found[0].Text {
		t.Fatalf("got %+v, %v", found, err)
	}
	digits, _ := FindAll(`\d{3}`, text, false)
	if len(digits) != 1 || digits[0].Text != cp(0xFF11, 0xFF12, 0xFF13) {
		t.Fatalf("got %+v", digits)
	}
}

func TestCaseSensitivityIsAnOptionAndASCIIOnly(t *testing.T) {
	if !mustMatch(t, `\bTODO\b`, "todo", false) || mustMatch(t, `\bTODO\b`, "todo", true) {
		t.Fatal("case option")
	}
	if mustMatch(t, `\bthis\b`, "th"+cp(0x0131)+"s", false) {
		t.Fatal("a dotless i is not an i")
	}
}

func TestAnAccentedLetterIsNotAWordCharacter(t *testing.T) {
	if !mustMatch(t, `\bcaf\b`, "caf"+cp(0x00E9), false) || mustMatch(t, `^\w+\s`, "caf"+cp(0x00E9)+" ok", false) {
		t.Fatal("accented letter")
	}
}

func TestUnlessOnlyCountsInsideTheSameSentence(t *testing.T) {
	rule := Rule{ID: "date", Patterns: []string{`\bready\s+by\s+friday\b`}, Unless: []string{`\bif\b`}}
	if got, _ := Evaluate(rule, "Ready by Friday if it passes."); len(got) != 0 {
		t.Fatalf("got %+v", got)
	}
	if got, _ := Evaluate(rule, "Ready by Friday. If you ask me, late."); len(got) != 1 {
		t.Fatalf("got %+v", got)
	}
}

func TestARangeFoundByTwoPatternsCountsOnce(t *testing.T) {
	if got, _ := Evaluate(Rule{ID: "twice", Patterns: []string{`\bharmless\b`, "harmless"}}, "harmless"); len(got) != 1 {
		t.Fatalf("got %+v", got)
	}
}

func TestFoldingKeepsAMapBackToTheOriginalText(t *testing.T) {
	text := "it" + cp(0x2019) + "s" + cp(0x000D, 0x2028) + "ok " + cp(0x2014) + " fine" + cp(0x000B) + "!"
	folded := Fold(text)
	if folded.Text != "it's\n\nok - fine !" {
		t.Fatalf("got %q", folded.Text)
	}
	if start, end := folded.OriginalRange(2, 3); text[start:end] != cp(0x2019) {
		t.Fatalf("the apostrophe maps to %q", text[start:end])
	}
}

func TestTheSentenceFunctionIncludesItsTerminator(t *testing.T) {
	text := "One. Two here! Three"
	if s, e := SentenceRange(text, 6); s != 5 || e != 14 {
		t.Fatalf("got %d %d", s, e)
	}
	if s, e := SentenceRange(text, 16); s != 15 || e != 20 {
		t.Fatalf("got %d %d", s, e)
	}
}

func TestPatternsEnginesReadDifferentlyAreRefused(t *testing.T) {
	backslash := string(rune(92))
	refused := []string{"(?=a)", "(?<!a)b", `(a)\1`, "(?<name>a)", "(?i)a", "a$", `\p{L}`, "a++", "a{,3}", "x{",
		"[[a]]", "[a&&b]", "[]a]", `\x41`, backslash + "u0041", `\Aa`, `\v`, "caf" + cp(0x00E9), `[\b]`, "a{1001}", "a{2,5000}"}
	for _, source := range refused {
		if len(CheckPattern(source)) == 0 {
			t.Errorf("accepted %q", source)
		}
	}
	accepted := []string{`\bcolou?r\b`, "a{2,3}", "a{2}", `[a-z\]]`, `\(\?=`, "(?:ab)+?", `\d+\.\d*`, "[^.!?]{0,30}", `\$5`, "a{1000}"}
	for _, source := range accepted {
		if problems := CheckPattern(source); len(problems) != 0 {
			t.Errorf("refused %q: %v", source, problems)
		}
	}
	if _, err := Compile("(?=a)", false); err == nil {
		t.Error("compiled a lookahead")
	}
	if _, err := Compile("(a", false); err == nil {
		t.Error("compiled an unbalanced group")
	}
}

func TestEveryVariantFoldsBackToTheOriginal(t *testing.T) {
	original := `It's "fine" - order #123456 kept`
	variants := Variants(original, false)
	var names []string
	for _, v := range variants {
		names = append(names, v.Name)
		// The Kelvin sign folds to a capital K, so compare without case.
		if strings.ToLower(Fold(v.Text).Text) != strings.ToLower(Fold(original).Text) {
			t.Errorf("%s does not fold back: %q", v.Name, Fold(v.Text).Text)
		}
	}
	want := []string{"long s", "Kelvin sign", "full-width", "zero-width spaces", "no-break spaces", "en dashes", "smart quotes"}
	if !reflect.DeepEqual(names, want) {
		t.Fatalf("got %v", names)
	}
}

// --- conformance: the same answers as conformance/*.tsv, which every other package also reads ---

func unescape(text string) string {
	var b strings.Builder
	for i := 0; i < len(text); i++ {
		if text[i] == 92 && i+5 < len(text) && text[i+1] == 'u' {
			code, err := strconv.ParseUint(text[i+2:i+6], 16, 32)
			if err == nil {
				b.WriteRune(rune(code))
				i += 5
				continue
			}
		}
		b.WriteByte(text[i])
	}
	return b.String()
}

func rows(t *testing.T, name string) [][]string {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "conformance", name))
	if err != nil {
		t.Fatal(err)
	}
	var out [][]string
	for _, line := range strings.Split(string(data), "\n") {
		if line != "" && !strings.HasPrefix(line, "#") {
			out = append(out, strings.Split(line, "\t"))
		}
	}
	return out
}

func conformanceRules(t *testing.T) map[string]*Rule {
	rules := map[string]*Rule{}
	for _, row := range rows(t, "rules.tsv") {
		rule, ok := rules[row[0]]
		if !ok {
			rule = &Rule{ID: row[0], CaseSensitive: row[1] == "sensitive"}
			rules[row[0]] = rule
		}
		if row[2] == "pattern" {
			rule.Patterns = append(rule.Patterns, unescape(row[3]))
		} else {
			rule.Unless = append(rule.Unless, unescape(row[3]))
		}
	}
	return rules
}

// Conformance positions count characters; Go reports bytes.
func characters(text string, byteOffset int) int { return utf8.RuneCountInString(text[:byteOffset]) }

func TestConformanceEveryCaseGetsTheRecordedAnswerAndFindings(t *testing.T) {
	rules := conformanceRules(t)
	cases := rows(t, "cases.tsv")
	if len(cases) < 100 {
		t.Fatalf("only %d cases", len(cases))
	}
	for _, row := range cases {
		text := unescape(row[3])
		found, err := Evaluate(*rules[row[0]], text)
		if err != nil {
			t.Fatal(err)
		}
		var spans []string
		for _, f := range found {
			spans = append(spans, strconv.Itoa(characters(text, f.Start))+":"+strconv.Itoa(characters(text, f.End)))
		}
		got := strings.Join(spans, ",")
		if got == "" {
			got = "-"
		}
		label := row[0] + " " + row[1] + " (" + row[2] + ")"
		if (len(found) > 0) != (row[1] == "match") || got != row[4] {
			t.Errorf("%s: got %s, want %s", label, got, row[4])
		}
	}
}

func TestConformanceVariantsAreGeneratedExactlyAsRecorded(t *testing.T) {
	rules := conformanceRules(t)
	var base []string
	var expected []Variant
	check := func() {
		if base == nil {
			return
		}
		got := Variants(unescape(base[3]), rules[base[0]].CaseSensitive)
		if len(got) == 0 && len(expected) == 0 {
			return
		}
		if !reflect.DeepEqual(got, expected) {
			t.Errorf("%s: got %q, want %q", base[3], got, expected)
		}
	}
	for _, row := range rows(t, "cases.tsv") {
		if row[2] == "as written" {
			check()
			base, expected = row, nil
		} else {
			expected = append(expected, Variant{Name: row[2], Text: unescape(row[3])})
		}
	}
	check()
}

func TestConformancePortabilityAgrees(t *testing.T) {
	patterns := rows(t, "patterns.tsv")
	if len(patterns) < 10 {
		t.Fatalf("only %d patterns", len(patterns))
	}
	for _, row := range patterns {
		source := unescape(row[1])
		refused := len(CheckPattern(source)) > 0
		if refused != (row[0] == "refused") {
			t.Errorf("%q: refused=%v, want %s", source, refused, row[0])
		}
	}
}
